### PR TITLE
[FLINK-21654][test] Added @throws information to tests

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

We had a few test failures that were supposedly caused by a bug in Hadoop Yarn 2.8.3 (see [YARN-7007](https://issues.apache.org/jira/browse/YARN-7007)). The related discussion can be found in [FLINK-15534](https://issues.apache.org/jira/browse/FLINK-15534). This PR tries to overcome this issue by catching the `NullPointerException` when calling `YarnClient.getApplications` and retrying the call.

## Brief change log

* Adds retry logic around `YarnClient.getApplications` and removes the recently added JavaDoc about the `NullPointerException`. Hopefully, this info is not necessary anymore considering the change.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
I tried reproducing the failure first in [this build](https://dev.azure.com/mapohl/flink/_build/results?buildId=331&view=results) with [4515bd2](https://github.com/XComp/flink/commit/4515bd2126ceaed8479202a9ce0c366e56d53bf4) without success. Looks like it's still happening not often enough? 🤔 

In a [second build](https://dev.azure.com/mapohl/flink/_build/results?buildId=332&view=logs&j=f3dc9b18-b77a-55c1-591e-264c46fe44d1&t=ebffacc3-9693-5a7f-11b7-68b343827cf3) I ran the `YARNSessionCapacitySchedulerITCase` in a loop with the workaround of [f6098d0c](https://github.com/apache/flink/commit/f6098d0c) without any failures, either.

**Comment by XComp** [comment]
I left the commits unsquashed for now to be transparent on the test runs as well.

File: flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java [hunk 1]
```diff
@@ -327,6 +328,36 @@ public void close() throws Exception {
...
+                return yarnClient.getApplications(states);
+            } catch (NullPointerException e) {
+                String npeStr = ExceptionUtils.stringifyException(e);
+                if (!npeStr.contains("RMAppAttemptMetrics.getAggregateAppResourceUsage")) {
+                    // unrelated NullPointerExceptions should be forwarded to the calling method
+                    throw e;
+                }
+
+                LOG.warn(
+                        "NullPointerException was caught most likely being related to YARN-7007. The related discussion is happening in FLINK-15534. The exception is going to be ignored.");
```
Comment (tillrohrmann) [inline flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java:351]: Maybe we could log `e` on debug if debug is enabled.

File: flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java [hunk 2]
```diff
@@ -327,6 +328,36 @@ public void close() throws Exception {
...
+            }
+        }
+
+        throw new IllegalStateException(
+                "YarnClient.getApplications command failed "
+                        + maxRetryCount
+                        + " times to gather the application report. Check FLINK-15534 for further details.");
```
Comment (tillrohrmann) [inline flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java:358]: Let's add `e` as the cause for the sake of completeness.

File: flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java [hunk 3]
```diff
@@ -327,6 +328,36 @@ public void close() throws Exception {
         }
     }
 
+    private static List<ApplicationReport> getApplicationReportWithRetryOnNPE(
+            final YarnClient yarnClient) throws IOException, YarnException {
+        return getApplicationReportWithRetryOnNPE(yarnClient, null);
+    }
+
+    private static List<ApplicationReport> getApplicationReportWithRetryOnNPE(
+            final YarnClient yarnClient, EnumSet<YarnApplicationState> states)
```
Comment (tillrohrmann) [inline flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java:337]:
```suggestion
            final YarnClient yarnClient, @Nullable EnumSet<YarnApplicationState> states)
```

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @XComp. I think the change makes sense. I had a few minor comments.

Why is this problem only limited to when we close the `YarnTestBase`? According to YARN-7007 it reads as if the failing of `YarnClient.getApplications` is not restricted to a specific state/timing?

**Comment by XComp** [comment]
> Thanks for creating this PR @XComp. I think the change makes sense. I had a few minor comments.
> 
> Why is this problem only limited to when we close the `YarnTestBase`? According to YARN-7007 it reads as if the failing of `YarnClient.getApplications` is not restricted to a specific state/timing?

Are you referring to the PR description here? You're right. It's not limited to the `close()` call. The description was out-dated. I realized during the implementation that `getApplication` gets also called in other situations. This is covered by the most recent changes. I updated the PR description accordingly.

**Comment by XComp** [comment]
I applied your comments, squashed the commits and rebased the branch.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating the PR @XComp. What about the other usages of `YarnClient.getApplications()`? I think we have a couple of other tests which also use this call? Are thy susceptible to the same problem? If yes, then we might wanna guard them as well.

**Comment by XComp** [comment]
Now I got your point. I just checked the `YarnTestBase`-internal calls but didn't check other classes. I updated the code accordingly.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @XComp. LGTM. Merging this once AZP gives green light.

**Pull request closed without merging**